modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java

```diff
@@ -0,0 +1,327 @@
+
+            int partitions = getPartitionCountFromCatalog(indexId, parameters.catalogVersion() - 1);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            return metaStorageManager.invoke(
+                    exists(startBuildIndexKey),
+                    concat(
+                            List.of(remove(startBuildIndexKey)),
+                            removePartitionBuildIndexOperations(indexId, partitions)
+                    ),
+                    List.of(noop())
+            );
+        });
+    }
+
+    private CompletableFuture<?> onIndexAvailable(MakeIndexAvailableEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            ByteArray startBuildIndexKey = startBuildIndexKey(parameters.indexId());
+
+            return metaStorageManager.invoke(exists(startBuildIndexKey), remove(startBuildIndexKey), noop());
+        });
+    }
+
+    private CompletableFuture<?> onUpdatePartitionBuildIndexKey(WatchEvent event) {
+        return inBusyLockAsync(busyLock, () -> {
+            if (!event.single()) {
+                // We don't need to handle keys on index creation or deletion.
+                return completedFuture(null);
+            }
+
+            Entry entry = event.entryEvent().newEntry();
+
+            if (!entry.tombstone()) {
+                // In case an index was created when there was only one partition.
+                return completedFuture(null);
+            }
+
+            String partitionBuildIndexKey = new String(entry.key(), UTF_8);
+
+            int indexId = parseIndexIdFromPartitionBuildIndexKey(partitionBuildIndexKey);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            long metastoreRevision = entry.revision();
+
+            if (isRemainingPartitionBuildIndexKeys(indexId, metastoreRevision)
+                    || isMetastoreKeyAbsent(startBuildIndexKey, metastoreRevision)) {
+                return completedFuture(null);
+            }
+
+            // We can use the latest version of the catalog since we are on the metastore thread.
+            CatalogIndexDescriptor indexDescriptor = getIndexDescriptorStrict(indexId, catalogManager.latestCatalogVersion());
+
+            // We will not wait for the command to be executed, since we will then find ourselves in a dead lock since we will not be able
+            // to free the metastore thread.
+            catalogManager.execute(buildMakeIndexAvailableCommand(indexDescriptor));
+
+            return completedFuture(null);
+        });
+    }
+
+    private void onIndexBuildForPartition(int indexId, int partitionId) {
+        inBusyLock(busyLock, () -> {
+            ByteArray partitionBuildIndexKey = partitionBuildIndexKey(indexId, partitionId);
+
+            // We don't need to wait for the operation to complete.
+            metaStorageManager.invoke(exists(partitionBuildIndexKey), remove(partitionBuildIndexKey), noop());
```

Comment:
What if the invocation fails (synchronously or asynchronously, inside a future)? Will partition 'under build' status remain forever (or till a restart)?

Reply:
Also, if it fails inside a future, we'll not see an exception. `whenComplete()` might be useful to log it.

Reply:
> What if the invocation fails (synchronously or asynchronously, inside a future)? Will partition 'under build' status remain forever (or till a restart)?

If I understand you correctly, then such a problem will have to be solved and taken into account in recovery ticket https://issues.apache.org/jira/browse/IGNITE-20637